control/pid.go

```diff
@@ -49,54 +75,139 @@ func (p *basicPID) GetTuning() bool {
+				p.tuner.tuning = false
+			}
+			p.y[0].SetSignalValueAt(0, out)
 		}
-		p.y[0].SetSignalValueAt(0, out)
 	} else {
-		dtS := dt.Seconds()
-		pvError := x[0].GetSignalValueAt(0)
-		p.int += p.kI * pvError * dtS
-		switch {
-		case p.int >= p.satLimUp:
-			p.int = p.satLimUp
-		case p.int <= p.satLimLo:
-			p.int = p.satLimLo
-		default:
-		}
-		deriv := (pvError - p.error) / dtS
-		output := p.kP*pvError + p.int + p.kD*deriv
-		p.error = pvError
-		if output > p.limUp {
-			output = p.limUp
-		} else if output < p.limLo {
-			output = p.limLo
+		// Multi Input/Output Implementation
+		if p.useMulti {
+			for i := 0; i < len(p.PIDSets); i++ {
+				output := calculateSignalValue(p, x, dt, i)
+				p.y[0].SetSignalValueAt(i, output)
+			}
+
+			// Single Input/Output Implementation
+		} else {
+			dtS := dt.Seconds()
+			pvError := x[0].GetSignalValueAt(0)
+			p.int += p.kI * pvError * dtS
+			switch {
+			case p.int >= p.satLimUp:
+				p.int = p.satLimUp
+			case p.int <= p.satLimLo:
+				p.int = p.satLimLo
+			default:
+			}
+			deriv := (pvError - p.error) / dtS
+			output := p.kP*pvError + p.int + p.kD*deriv
+			p.error = pvError
+			if output > p.limUp {
+				output = p.limUp
+			} else if output < p.limLo {
+				output = p.limLo
+			}
+			p.y[0].SetSignalValueAt(0, output)
 		}
-		p.y[0].SetSignalValueAt(0, output)
 	}
 	return p.y, true
 }
 
+// For a given signal, compute new signal value based on current signal value, & its respective error.
+func calculateSignalValue(p *basicPID, x []*Signal, dt time.Duration, sIndex int) float64 {
+	dtS := dt.Seconds()
+	pvError := x[0].GetSignalValueAt(sIndex)
+	p.PIDSets[sIndex].int += p.PIDSets[sIndex].I * pvError * dtS
+
+	switch {
+	case p.PIDSets[sIndex].int >= p.satLimUp:
+		p.PIDSets[sIndex].int = p.satLimUp
+	case p.PIDSets[sIndex].int <= p.satLimLo:
+		p.PIDSets[sIndex].int = p.satLimLo
+	default:
+	}
+	deriv := (pvError - p.PIDSets[sIndex].signalErr) / dtS
+	output := p.PIDSets[sIndex].P*pvError + p.PIDSets[sIndex].int + p.PIDSets[sIndex].D*deriv
+	p.PIDSets[sIndex].signalErr = pvError
+	if output > p.limUp {
+		output = p.limUp
+	} else if output < p.limLo {
+		output = p.limLo
+	}
+
+	return output
+}
+
 func (p *basicPID) reset() error {
+	var ok bool
 	p.int = 0
 	p.error = 0
 
+	// Each PIDSet is taken from the config, if the attribute exists (it's optional).
+	// If PID Sets was given as an attribute, we know we're in 'multi' mode. For each
+	// set of PIDs we initialize its values to 0 and create a tuner object.
+	if p.cfg.Attribute.Has("PIDSets") {
+		p.PIDSets, ok = p.cfg.Attribute["PIDSets"].([]*PIDConfig)
+		if !ok {
+			return errors.New("PIDSet did not initialize correctly")
+		}
+		if len(p.PIDSets) > 0 {
+			p.useMulti = true
+			p.tuners = make([]*pidTuner, len(p.PIDSets))
+
+			for i := 0; i < len(p.PIDSets); i++ {
+				p.PIDSets[i].int = 0
+				p.PIDSets[i].signalErr = 0
+			}
```

Comment:
okay so in reference to my above comment about mismatch in number of pids and number of tuners, it seems like that comment is just wrong, because we're making a tuner for every pid block right?

Reply:
yeah I guess the comment is abit misleading. we make a structure that holds up to N tuners, based on the number of pid sets/gains/signals configured. 

a tuner only gets initialized if a pid set needs tuning, otherwise the tuning boolean will be set to `false`

I'll update the comment in getTuning to reflect this behavior